modules/plugin/mongodb/src/main/java/org/geotools/data/mongodb/AbstractFilterToMongo.java

```diff
@@ -520,27 +548,97 @@ public Object visit(Disjoint filter, Object extraData) {
 
-        DBObject geometryDBObject = geometryBuilder.toObject(geometry);
-        DBObject dbo =
-                BasicDBObjectBuilder.start()
-                        .push("$near")
-                        .add("$geometry", geometryDBObject)
-                        .add("$maxDistance", distanceInMeters)
-                        .get();
+        CoordinateReferenceSystem coordinateReferenceSystem =
+                featureType.getCoordinateReferenceSystem();
+        double distanceInMeters = DistanceBufferUtil.getDistanceInMeters(filter);
+        DBObject dbo;
+        // If point use $near
+        if (geometry instanceof Point) {
+            DBObject geometryDBObject = geometryBuilder.toObject(geometry);
+            dbo =
+                    BasicDBObjectBuilder.start()
+                            .push("$near")
+                            .add("$geometry", geometryDBObject)
+                            .add("$maxDistance", distanceInMeters)
+                            .get();
+            // Other type of geometry, use $geoIntersects
+        } else {
+            try {
+                Point centroid = geometry.getCentroid();
+                CoordinateReferenceSystem azimuthalEquidistantCRS =
+                        computeAzimutalEquidistantCRS(centroid, coordinateReferenceSystem);
+                // Transform geometry to CRS that supports meters
+                MathTransform transform =
+                        CRS.findMathTransform(coordinateReferenceSystem, azimuthalEquidistantCRS);
+                geometry = JTS.transform(geometry, transform);
+
+                logAdditionalInformation(geometry, JTS.transform(centroid, transform));
+
+                // Apply buffer in meters to geometry
+                Geometry bufferedGeometry = geometry.buffer(distanceInMeters, DEFAULT_SEGMENTS);
+
+                // Transform back to original with buffer applied
+                transform =
+                        CRS.findMathTransform(azimuthalEquidistantCRS, coordinateReferenceSystem);
+                geometry = JTS.transform(bufferedGeometry, transform);
+            } catch (FactoryException | TransformException e) {
+                throw new RuntimeException(
+                        "Failed to compute polygon within distance from reference geometry", e);
+            }
+            DBObject geometryDBObject = geometryBuilder.toObject(geometry);
+            dbo =
+                    BasicDBObjectBuilder.start()
+                            .push("$geoIntersects")
+                            .add("$geometry", geometryDBObject)
+                            .get();
+        }
 
         output.put((String) e1, dbo);
         return output;
     }
 
+    private static void logAdditionalInformation(Geometry geometry, Geometry transformedCentroid) {
+        if (geometry.intersects(transformedCentroid.buffer(HUNDRED_KM_IN_METERS).getBoundary())) {
```

Comment:
Sigh no... if the dwithin distance is > 100KM then log the warning.

Reply:
Wait .. hum.. actually... both! Your observation is right, the distances are preserved properly within 100km from the center of the projection. So it might be that the geometry is too big, or the distance is too big. But I'm not sure about the test... it would probably be easier if you take the buffered geometry, get its bbox, and compute the distance between one of its corners and the centroid. If > 100km, then we went too far.

Reply:
updated accordingly

Reply:
Changed to checking both distance from centroid to corners of envelope and input buffer size.